src/net.cpp

```diff
@@ -674,6 +674,18 @@ int V1TransportDeserializer::readData(const char *pch, unsigned int nBytes)
     memcpy(&vRecv[nDataPos], pch, nCopy);
     nDataPos += nCopy;
 
+    if (Complete()) {
```

Comment:
Is there a particular reason to check `Complete()` and do this here instead of in the complete check in `ReceiveMsgBytes()`?

Reply:
In the future, I'd like to see `TransportDeserializer` have a stateless interface such that it only produces valid messages or otherwise returns an appropriate error; see https://github.com/bitcoin/bitcoin/pull/16202#issuecomment-535152710. Not sure if @jonasschnelli agrees with me.

In that world, all validity checks would remain within the deserializer. Then `CNetMessage` would only need to represent valid messages so all its `m_valid_`* fields could be removed.

Reply:
@jkczyz I don't think that it's possible to have a stateless interface for `TransportDeserializer`. When we call into `Read(pch, nBytes)` the bytes that we've read from the socket and are passing in may not constitute the full message. That's why `TransportDeserializer` is stateful and keeps track of partially received headers/messages in `hdrbuf`/`vRecv`.

I agree with @narula that the error-handing should be done in `ReceiveMsgBytes()`. That's partially because it's neater: `Read()`, `readHeader()` and `readData()` could all be changed to return an error enum or object, and error-handling can all happen in one place. I also think that logging needs to happen within `CNode` for it to be meaningful. Without logging the node id, this log message is pretty useless.

Reply:
> @jkczyz I don't think that it's possible to have a stateless interface for `TransportDeserializer`. When we call into `Read(pch, nBytes)` the bytes that we've read from the socket and are passing in may not constitute the full message. That's why `TransportDeserializer` is stateful and keeps track of partially received headers/messages in `hdrbuf`/`vRecv`.
> 

IIRC, my linked comment was alluding to the fact that additional refactoring may be needed to accomplish this. For instance, decoupling reading bytes from the wire from deserializing those read bytes.